source/azure_iot_hub_client.c

```diff
@@ -346,6 +355,95 @@ static uint32_t prvGetTimeMs( void )
 }
 /*-----------------------------------------------------------*/
 
+static AzureIoTHubClientResult_t prvRequestIdGet( AzureIoTHubClient_t * pxAzureIoTHubClient,
+                                                  az_span xSpan,
+                                                  bool xOddSeq,
+                                                  uint32_t * pulRequestId,
+                                                  az_span * pxSpan )
+{
+    AzureIoTHubClientResult_t xReturn;
+    az_span xRemainder;
+    az_result xCoreResult;
+
+    if( xOddSeq == ( bool ) ( pxAzureIoTHubClient->_internal.ulCurrentRequestID & 0x01 ) )
+    {
+        pxAzureIoTHubClient->_internal.ulCurrentRequestID += 2;
+    }
+    else
+    {
+        pxAzureIoTHubClient->_internal.ulCurrentRequestID += 1;
+    }
+
+    if( pxAzureIoTHubClient->_internal.ulCurrentRequestID == 0 )
+    {
+        pxAzureIoTHubClient->_internal.ulCurrentRequestID = 2;
+    }
+
+    xCoreResult = az_span_u32toa( xSpan, pxAzureIoTHubClient->_internal.ulCurrentRequestID, &xRemainder );
+
+    if( az_result_failed( xCoreResult ) )
+    {
+        AZLogError( ( "Couldn't convert request id to span" ) );
+        xReturn = eAzureIoTHubClientFailed;
+    }
+    else
+    {
+        *pxSpan = az_span_slice( *pxSpan, 0, az_span_size( *pxSpan ) - az_span_size( xRemainder ) );
```

Comment:
Any [in] paramerter of span generally is passed as copy and out is pointer, which is waste of stack space in [inout] parameter. Few api's of span and their usage:

API: [https://github.com/Azure/azure-sdk-for-c/blob/5d8baab842e36e277aa58fdb7ad1f287f1375e4d/sdk/inc/azure/core/az_span.h#L412](https://github.com/Azure/azure-sdk-for-c/blob/5d8baab842e36e277aa58fdb7ad1f287f1375e4d/sdk/inc/azure/core/az_span.h#L412)
usage:[https://github.com/Azure/azure-sdk-for-c/blob/6a61fdc180cb43d9bb6e22d06166724bd4156e73/sdk/samples/iot/aziot_esp8266/aziot_esp8266.ino#L283](https://github.com/Azure/azure-sdk-for-c/blob/6a61fdc180cb43d9bb6e22d06166724bd4156e73/sdk/samples/iot/aziot_esp8266/aziot_esp8266.ino#L283)


---
In reply to: [612734661](https://github.com/Azure/azure-iot-middleware-freertos/pull/75#discussion_r612734661) [](ancestors = 612734661)

Reply:
We do not expose span and all usage are internal.
I will keep it same for now.... until we think otherwise.

---
In reply to: [612817618](https://github.com/Azure/azure-iot-middleware-freertos/pull/75#discussion_r612817618) [](ancestors = 612817618)